statsd/uds_blocking.go

```diff
@@ -0,0 +1,65 @@
+package statsd
+
+import (
+	"net"
+	"sync"
+	"time"
+)
+
+// blockingUdsWriter is an internal class wrapping around management of UDS connection
+type blockingUdsWriter struct {
+	// Address to send metrics to, needed to allow reconnection on error
+	addr net.Addr
+	// Established connection object, or nil if not connected yet
+	conn net.Conn
+	// write timeout
+	writeTimeout time.Duration
+	sync.Mutex   // used to lock conn / writer can replace it
+}
+
+// New returns a pointer to a new blockingUdsWriter given a socket file path as addr.
+func newBlockingUdsWriter(addr string) (*blockingUdsWriter, error) {
```

Comment:
I'm not sure we need a specific implementation for a blocking writer, I believe in effect the async writer with a datagram queue depth of 1 is essentially the same thing. There might be some details we want to look at - channel blocking vs mutex locking, etc, but I think it'd behave exactly as expected, it would perhaps reduce the amount of duped code here by thinking about that.

Reply:
The issue is error checking. Unlike UDP, UDS will error out if the agent does not receive the payload. With the blocking implementation you would get alerted if your message did not reach the agent.

However, in my opinion it's pretty flawed especially if the client buffers. In this case you only get one error on the call that actually sends the payload.

I think the best path might be:
- Merge this and #82
- (Optional) Set the blocking implementation as default
- Release 2.x
- Remove the blocking implementation
- Release 3.0

WDYT ? 
cc @xvello you might have an opinion